src/cc/bcc_elf.c

```diff
@@ -629,102 +649,107 @@ static char *find_debug_via_symfs(Elf *e, const char* path) {
+                                 NULL);
+  if (fd >= 0) {
+    if (bcc_elf_file_open_fd(fd, out)) {
+      close(fd);
+      fd = -1;
+    }
+  }
 
   debuginfod_end(client);
-  return debugpath;
+  return fd >= 0 ? 0 : -1;
 }
 #endif
 
-static char *find_debug_file(Elf* e, const char* path, int check_crc) {
-  char *debug_file = NULL;
-
-  // If there is a separate debuginfo file, try to locate and read it, first
-  // using symfs, then using the build-id section, finally using the debuglink
-  // section. These rules are what perf and gdb follow.
-  // See:
-  // - https://github.com/torvalds/linux/blob/v5.2/tools/perf/Documentation/perf-report.txt#L325
-  // - https://sourceware.org/gdb/onlinedocs/gdb/Separate-Debug-Files.html
-  debug_file = find_debug_via_symfs(e, path);
-  if (!debug_file)
-    debug_file = find_debug_via_buildid(e);
-  if (!debug_file)
-    debug_file = find_debug_via_debuglink(e, path, check_crc);
+// Returns 0 on success, otherwise nonzero.
+// If successfull, 'out' param is a valid bcc_elf_file.
+// Caller is responsible for calling bcc_elf_file_close when done using it.
+static int find_debug_file(Elf *e, const char *path, int check_crc,
+                           struct bcc_elf_file *out) {
+  if (find_debug_via_symfs(e, path, out) == 0)
+    return 0;
+
+  if (find_debug_via_buildid(e, out) == 0)
+    return 0;
+
+  if (find_debug_via_debuglink(e, path, check_crc, out) == 0)
+    return 0;
+
 #ifdef HAVE_LIBDEBUGINFOD
-  if (!debug_file)
-    debug_file = find_debug_via_debuginfod(e);
+  if (find_debug_via_debuginfod(e, out) == 0)
+    return 0;
 #endif
 
-  return debug_file;
+  return -1;
 }
 
 static int foreach_sym_core(const char *path, bcc_elf_symcb callback,
                             bcc_elf_symcb_lazy callback_lazy,
-                            struct bcc_symbol_option *option, void *payload,
-                            int is_debug_file) {
-  int res;
-  char *debug_file;
+                            struct bcc_symbol_option *option, void *payload) {
   struct bcc_elf_file elf_file;
   bcc_elf_file_init(&elf_file);
+  struct bcc_elf_file debug_elf_file;
+  bcc_elf_file_init(&debug_elf_file);
+  int res;
 
   if (!option)
     return -1;
 
   if (bcc_elf_file_open(path, &elf_file) < 0)
     return -1;
 
-  if (option->use_debug_file && !is_debug_file) {
-    // The is_debug_file argument helps avoid infinitely resolving debuginfo
-    // files for debuginfo files and so on.
-    debug_file =
-        find_debug_file(elf_file.elf, path, option->check_debug_file_crc);
-    if (debug_file) {
-      foreach_sym_core(debug_file, callback, callback_lazy, option, payload, 1);
-      free(debug_file);
+  if (option->use_debug_file) {
+    if (find_debug_file(elf_file.elf, path, option->check_debug_file_crc,
+                        &debug_elf_file) == 0) {
+      listsymbols(debug_elf_file.elf, callback, callback_lazy, payload, option,
+                  1);
+      bcc_elf_file_close(&debug_elf_file);
```

Comment:
Return ?

Reply:
That's not how it used to work. I am worried that some symbols might be defined only in the non-debug file.

Reply:
OK, I misread this.